### PR TITLE
job model design

### PR DESCRIPTION
**Pull request by 8olio8**

### CODE REVIEW

File: pkg/common/schema/job.go [hunk 1]
```diff
@@ -20,11 +20,15 @@ import (
...
 	"encoding/json"
 	"fmt"
+
+	v1 "k8s.io/api/core/v1"
 )
 
 type JobType string
 type ActionOnJob string
 type JobStatus string
+type PFJobType string
```
Comment (D0m021ng) [inline pkg/common/schema/job.go]: PFJobType is similar with JobType

File: pkg/common/schema/job.go [hunk 2]
```diff
@@ -154,10 +168,60 @@ type PFJobConf interface {
...
+}
+
+type FileSystem struct {
+	Name      string `json:"name"`
+	MountPath string `json:"mountPath,omitempty"`
+	SubPath   string `json:"subPath,omitempty"`
+	ReadOnly  bool   `json:"readOnly,omitempty"`
+}
+
+type JobFlavour struct {
```
Comment (D0m021ng) [inline pkg/common/schema/job.go]: Duplicate definition, `pkg/common/schema/flavor.go` defines `Flavour` struct.

File: pkg/apiserver/models/job_lable.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package models
```
Comment (D0m021ng) [inline pkg/apiserver/models/job_lable.go]: Typo: `job_lable.go` =>` job_label.go`

File: pkg/apiserver/models/job_lable.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"time"
+
+	"gorm.io/gorm"
+)
+
+type JobLabel struct {
+	Pk        int64          `json:"-" gorm:"primaryKey;autoIncrement"`
+	Label     string         `json:"label" gorm:"type:varchar(255);NOT NULL"`
```
Comment (D0m021ng) [inline pkg/apiserver/models/job_lable.go]: Add a field `id` for job_label table ?

File: pkg/common/schema/job.go [hunk 5]
```diff
@@ -154,10 +168,60 @@ type PFJobConf interface {
...
+}
+
+type StatusInfo struct {
+	Name       string            `json:"name"`
+	ID         string            `json:"id"`
+	Status     v1.PodStatus      `json:"status"`
+	Containers []ContainerStatus `json:"containers"`
+}
+
+type ContainerStatus struct {
```
Comment (D0m021ng) [inline pkg/common/schema/job.go]: Remove ContainerStatus

File: pkg/common/schema/job.go [hunk 6]
```diff
@@ -154,10 +168,60 @@ type PFJobConf interface {
...
+	ID       string `json:"id"`
+	Replicas int    `json:"replicas"`
+	Role     string `json:"role"`
+	Conf
+}
+
+type StatusInfo struct {
+	Name       string            `json:"name"`
+	ID         string            `json:"id"`
+	Status     v1.PodStatus      `json:"status"`
```
Comment (D0m021ng) [inline pkg/common/schema/job.go]: interface{} ?

File: pkg/common/schema/resource.go [hunk 7]
```diff
@@ -28,12 +28,12 @@ type ResourceName string
 
 // Resource struct defines all the resource type
 type Resource struct {
-	MilliCPU float64
-	Memory   float64
+	MilliCPU float64 `json:"cpu"`
+	Memory   float64 `json:"memory"`
 	Storage  float64
```
Comment (D0m021ng) [inline pkg/common/schema/resource.go]: `json:"-"`

File: pkg/apiserver/models/job.go [hunk 8]
```diff
@@ -31,20 +31,33 @@ import (
...
+	Config            schema.Conf      `json:"config" gorm:"type:text"`
+	RuntimeInfoJson   string           `json:"-" gorm:"column:runtime_info;default:'{}'"`
+	RuntimeInfo       interface{}      `json:"runtimeInfo" gorm:"-"`
+	Status            schema.JobStatus `json:"status"`
+	Message           string           `json:"message"`
+	ResourceJson      string           `json:"-" gorm:"type:text;default:'{}'""`
+	Resource          *schema.Resource `json:"resource" gorm:"-"`
+	Framework         schema.Framework `json:"framework" gorm:"type:varchar(30)"`
+	Members           []Member         `json:"members" gorm:"type:text"`
+	ExtensionTemplate string           `json:"extensionTemplate" gorm:"type:text"`
```
Comment (qiaoshuangshuang) [inline pkg/apiserver/models/job.go:48]: 考虑数据库升级的方案了吗，新增字段可以为空么

File: pkg/apiserver/models/job.go [hunk 9]
```diff
@@ -31,20 +31,33 @@ import (
...
-	RuntimeInfo     interface{}      `json:"runtimeInfo" gorm:"-"`
-	Status          schema.JobStatus `json:"status"`
-	Message         string           `json:"message"`
-	CreatedAt       time.Time        `json:"createTime"`
-	ActivatedAt     sql.NullTime     `json:"activateTime"`
-	UpdatedAt       time.Time        `json:"updateTime,omitempty"`
-	DeletedAt       gorm.DeletedAt   `json:"-" gorm:"index"`
+	Pk                int64            `json:"-" gorm:"primaryKey;autoIncrement"`
+	ID                string           `json:"jobID" gorm:"type:varchar(60);uniqueIndex"`
+	UserName          string           `json:"userName" gorm:"type:varchar(255);NOT NULL"`
```
Comment (qiaoshuangshuang) [inline pkg/apiserver/models/job.go:36]: sql文件也要提交上来

File: pkg/apiserver/models/job.go [hunk 10]
```diff
@@ -31,20 +31,33 @@ import (
...
+	CreatedAt         time.Time        `json:"createTime"`
+	ActivatedAt       sql.NullTime     `json:"activateTime"`
+	UpdatedAt         time.Time        `json:"updateTime,omitempty"`
+	DeletedAt         gorm.DeletedAt   `json:"-" gorm:"index"`
+}
+
+type Member struct {
+	ID          string `json:"id"`
+	Replicas    int    `json:"replicas"`
+	Role        string `json:"role"`
```
Comment (qiaoshuangshuang) [inline pkg/apiserver/models/job.go]: role为啥不是枚举

File: pkg/apiserver/models/job_label.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+type JobLabel struct {
+	Pk        int64  `gorm:"primaryKey;autoIncrement"`
+	ID        string `gorm:"type:varchar(36);uniqueIndex"`
+	Label     string `gorm:"type:varchar(255);NOT NULL"`
+	JobID     string `gorm:"type:varchar(60);NOT NULL"`
+	CreatedAt time.Time
+	DeletedAt gorm.DeletedAt `gorm:"index"`
+}
+
+func (JobLabel) TableName() string {
```
Comment (qiaoshuangshuang) [inline pkg/apiserver/models/job_label.go:18]: job的labels单独做了表，那annotation为啥不需要

Comment (D0m021ng) [reply pkg/apiserver/models/job_label.go:18]: In order to query job by multi dimensional labels, such as query job by `org_id=xx` and `project_id=xx` .

File: pkg/apiserver/models/job.go (hunk 8 above)
Comment (D0m021ng) [reply pkg/apiserver/models/job.go:48]: The new field can be empty.

Comment (8olio8) [reply pkg/apiserver/models/job.go:48]: 可以为空

File: pkg/apiserver/models/job_label.go (hunk 11 above)
Comment (8olio8) [reply pkg/apiserver/models/job_label.go:18]: 考虑到目前list接口中暂时只支持对label的过滤，所以目前只有labels单做了表

File: pkg/common/schema/job.go [hunk 12]
```diff
@@ -80,7 +83,22 @@ const (
 	StatusJobTerminating JobStatus = "terminating"
 	StatusJobTerminated  JobStatus = "terminated"
 	StatusJobCancelled   JobStatus = "cancelled"
-	StatusJobSkipped   JobStatus = "skipped"
+	StatusJobSkipped     JobStatus = "skipped"
+
+	RoleMaster  RoleMember = "master"
```
Comment (qiaoshuangshuang) [inline pkg/common/schema/job.go]: i cannot see the define of the spark app

**Review by qiaoshuangshuang — APPROVED** [review]

**Pull request merged**